server/bin/pbench-audit-server.py

```diff
@@ -0,0 +1,581 @@
+                if hid_entries_path.is_file():
+                    hier.add_unexpected_entries(
+                        Hierarchy.UNEXPECTED_OBJECTS, controller, hid_entries_path.name,
+                    )
+        controller_subdir = list()
+        for item in direct_entries:
+            item_path = Path(item)
+            if item_path.is_dir():
+                controller_subdir.append(item_path.name)
+            elif item_path.is_symlink():
+                try:
+                    item_p = item_path.resolve(strict=True)
+                except FileNotFoundError:
+                    hier.add_error_or_inconceivable_entries(
+                        Hierarchy.ERROR,
+                        controller,
+                        f"{item_path.name}_path, '{item_path}', does not resolve to a real location",
+                    )
+                symlink_item = f"{item_path.name} -> {item_p}"
+                hier.add_unexpected_entries(
+                    Hierarchy.UNEXPECTED_SYMLINKS, controller, symlink_item
+                )
+            elif all(
+                [
+                    item_path.is_file(),
+                    not item_path.name.endswith(".tar.xz"),
+                    not item_path.name.endswith(".tar.xz.md5"),
+                ]
+            ):
+                hier.add_unexpected_entries(
+                    Hierarchy.UNEXPECTED_OBJECTS, controller, item_path.name
+                )
+            elif item_path.is_file() and (
+                item_path.name.endswith(".tar.xz")
+                or item_path.name.endswith(".tar.xz.md5")
+            ):
+                hier.add_tarballs(controller)
+            else:
+                hier.add_error_or_inconceivable_entries(
+                    Hierarchy.INCONCEIVABLE, controller, item_path.name,
+                )
+        verify_subdirs(hier, controller, controller_subdir)
+        verify_prefixes(hier, controller)
+
+    return 0
+
+
+# Incoming Hierarchy
+def verify_tar_dirs(ihier, tarball_dirs, tblist, controller):
+    for tb in tarball_dirs:
+        if tb.endswith("unpack"):
+            tar = tb[:-7]
+            tar = f"{tar}.tar.xz"
+            dict_val = Hierarchy.INVALID_UNPACKING_DIRS
+        else:
+            tar = f"{tb}.tar.xz"
+            dict_val = Hierarchy.INVALID_TB_DIRS
+        tarfile = Path(ihier.config.ARCHIVE, controller, tar)
+        if tarfile.exists():
+            if os.access(tarfile, os.R_OK):
+                continue
+        else:
+            tblist(dict_val, controller, tb)
+    return 0
+
+
+# 'ihier' is IncomingHierarchy class object
+def verify_incoming(ihier, verifylist):
+
+    for controller in verifylist:
+        ihier.add_controller(controller)
+
+        if not Path(ihier.config.ARCHIVE, controller).is_dir():
+            # Skip incoming controller directories that don't have an $ARCHIVE
+            # directory, handled in another part of the audit.
+            continue
```

Comment:
This feels weird to me, is it correct?  If `controller` is not a directory (in the archive tree), do we want to add it at line 186?  Or, should the order of these be swapped?

Reply:
yes, we do. That is more for a purpose for keeping track of all controllers. on the other hand we dont want to process anything inside it.